### ISSUE TITLE
Invalid Auth Token on every request

### ISSUE THREAD

**Issue by anishzute** [timkpaine/tdameritrade#87]
After having saved the `client_id` and `redirect_uri` as env variables and calling them, I am able to successfully run `tdameritrade.auth.authentication()` and get an access token as follows:

` >>> tdameritrade.auth.authentication(client_id,redirect_uri,os.getenv('TDAUSER'), os.getenv('TDAPASS'))`

result:

`{'access_token': 'long_access_token_string', 'scope': 'PlaceTrades AccountAccess MoveMoney', 'expires_in': 1800, 'refresh_token_expires_in': 7776000, 'token_type': 'Bearer'}`

I then initialize a client using `td = tdameritrade.TDClient()`

if I go to request a quote using `td.quote('SPY')` however, I get the following error:

`Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/anishzute/anaconda3/envs/OptionsProfitCalc/lib/python3.8/site-packages/tdameritrade/client.py", line 108, in quote
    resp = self._request(QUOTES,
  File "/home/anishzute/anaconda3/envs/OptionsProfitCalc/lib/python3.8/site-packages/tdameritrade/client.py", line 27, in _request
    handle_error_response(resp)
  File "/home/anishzute/anaconda3/envs/OptionsProfitCalc/lib/python3.8/site-packages/tdameritrade/exceptions.py", line 12, in handle_error_response
    raise codes[resp.status_code]()
tdameritrade.exceptions.InvalidAuthToken: -1: An unknown error occurred`

What's weird is that this was working for me just last week with no change in the code. Any ideas?

**Comment by nmcginty** [comment]
If you take a look in your result dictionary you will notice the 'expires_in' key has a value of 1800 seconds (30 minutes). Thus, the Auth token is only valid for 30 minutes. Calls to the API with an invalid Auth token will produce the error you described. If this Auth token was requested a week ago then it is surely invalid.

You can request a new one making a call with your refresh token and consumer key. 
`
import request
REFRESH_TOKEN = # fill in this 
CONSUMER_KEY = # fill in this
ACCESS_TOKEN_ENDPOINT= "https://api.tdameritrade.com/v1/oauth2/token"

payload = {  'grant_type': 'refresh_token',
                    'refresh_token': REFRESH_TOKEN,
                    'client_id': CONSUMER_KEY }

response = request.post(url=ACCESS_TOKEN_ENDPOINT, data=payload, verify=True)
access_token = response['access_token']
`
Of course this will not update the **_token** field on the TD Client or session object. 
An alternative is running `tdameritrade.auth.authentication(client_id,redirect_uri,os.getenv('TDAUSER'), os.getenv('TDAPASS'))` prior to making any calls to the API, this will request a new token for you with a time to live of 30 minutes.

**Comment by dahifi** [comment]
The new session handler will automatically refresh the access token if it gets a 401 error. Once you’ve gotten the token via the chrome driver, add it to your ENV. 

Refresh tokens are valid for several months, so you shouldn’t see the auth process for a long time once you’ve added it.

**Comment by dahifi** [comment]
The new session handler will automatically refresh the access token if it gets a 401 error. Once you’ve gotten the token via the chrome driver, add it to your ENV. 

Refresh tokens are valid for several months, so you shouldn’t see the auth process for a long time once you’ve added it. 

I'm currently using the following in one of my projects: 

```
import os
from tdameritrade.client import TDClient

client_id = os.environ['AMERITRADE_CLIENT_ID']
refresh_token = os.environ['AMERITRADE_REFRESH_TOKEN']

tdclient = TDClient(client_id=client_id, refresh_token=refresh_token)

def get_df(symbol):
    return tdclient.historyDF(symbol, periodType="year", frequencyType="daily")
```

**Comment by anishzute** [comment]
Thanks for the responses! I think I was a bit unclear: I am trying to authenticate and get a new access token every time I run the script, yet I still am unable to. Today, I got a different, weird error. This is the code I am running, from top to bottom. As you can see, I am trying to authenticate using my client_id, redirect_url, and username and password. It goes throw the process of opening up a chromium browser, authenticating with the TDAmeritrade website, and closing out the browser. I even seem to get a valid response in the following JSON form: `{'access_token': 'access_token_string', 'refresh_token': 'refresh_token_string', 'scope': 'PlaceTrades AccountAccess MoveMoney', 'expires_in': 1800, 'refresh_token_expires_in': 7776000, 'token_type': 'Bearer'}`

This is my code:

```
import tdameritrade, os
from tdameritrade import auth

client_id = os.getenv('TDAMERITRADE_CLIENT_ID')
redirect_uri = os.getenv('REDIRECT_URL')

tdameritrade.auth.authentication(client_id,redirect_uri,os.getenv('TDAUSER'),os.getenv('TDAPASS'))
td = tdameritrade.TDClient()
```
This runs with no errors.

However, when I try to get the quote, using `td.quote('SPY')`,  I'm getting the error:

```
  File "/home/anishzute/anaconda3/envs/OptionsProfitCalc/lib/python3.7/os.py", line 681, in __getitem__
    raise KeyError(key) from None
KeyError: 'ACCESS_TOKEN'
```

Maybe I'm doing this wrong and somehow the access token that came from the `authentication` request isn't being passed to the `quote` request?

**Comment by dahifi** [comment]
You are correct, nothing is passed from the authentication request to the TDClient, it's just returned as JSON and it's up to you what to do with it from there. 

Please note the constructor for the [TDClient class](https://github.com/timkpaine/tdameritrade/blob/a0330b10e0cb02f23bc68645e8bc9577fd3987e5/tdameritrade/client.py#L76):

```
class TDClient(object):
    def __init__(self, client_id=None, refresh_token=None, account_ids=None):
        self._clientId = client_id or os.environ['TDAMERITRADE_CLIENT_ID']
        self._refreshToken = refresh_token or os.environ['TDAMERITRADE_REFRESH_TOKEN']
        self.accountIds = account_ids or []
        self.session = TDASession(self._refreshToken, self._clientId)
```
At a minimum, the client id and refresh token need to be passed explicitly or loaded in the environment.

**Comment by anishzute** [comment]
I figured out the problem.

This is the constructor I am seeing. I pip installed the package. Is this a deprecated version?

```
class TDClient(object):
    def __init__(self, access_token=None, accountIds=None):
        self._token = access_token or os.environ['ACCESS_TOKEN']
        self.accountIds = accountIds or []
        self.session = TDASession()
        if self._token:
            self.session.set_token(self._token)
```

Using this info, I first set the `ACCESS_TOKEN` var manually. The following code works: 

```
auth = tdameritrade.auth.authentication(client_id,redirect_uri,os.getenv('TDAUSER'), os.getenv('TDAPASS'))

os.environ['TDAMERITRADE_REFRESH_TOKEN'] = auth['refresh_token']
os.environ['ACCESS_TOKEN'] = auth['access_token']
td = tdameritrade.TDClient()

print(td.quote('SPY'))
````

**Comment by dahifi** [comment]
Yes, that's an older version. @timkpaine, you know more about the package management than I do. Do we need a version bump? 

Ani, the new version will just take the refresh token. Everytime you authenticate using the full call, you'll get a new set of tokens. The refresh token can be shared across projects and will get you a new access token without having to go through the full process.

**Comment by roysta2004** [comment]
Yes even tho the docs have been updated to reflect the new refresh token and refers to this being in version 0.1, 0.1 installed via pip does not include the new code.  I downloaded the latest files and installed manually in order to use the new authentication mentioned in the docs with the refresh token.  (download and extract the zip and run pip install with the '-e' flag from the directory containing the setup.py file.
